core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,280 @@ class DAGSchedulerSuite
+
+  /**
+   * In this test we simulate a job failure where there are two failures in two different stages.
+   * Specifically, stage1 fails twice, and then stage2 twice. In total, the job has had four
+   * failures overall but not four failures for a particular stage, and as such should not be
+   * aborted.
+   */
+  test("Failures in different stages should not trigger an overall abort") {
+    setupStageAbortTest(sc)
+
+    val shuffleOneRdd = new MyRDD(sc, 2, Nil).cache()
+    val shuffleDepOne = new ShuffleDependency(shuffleOneRdd, null)
+    val shuffleTwoRdd = new MyRDD(sc, 2, List(shuffleDepOne)).cache()
+    val shuffleDepTwo = new ShuffleDependency(shuffleTwoRdd, null)
+    val finalRdd = new MyRDD(sc, 1, List(shuffleDepTwo))
+    submit(finalRdd, Array(0))
+
+    // In the first two iterations, Stage 0 succeeds and stage 1 fails. In the next two iterations,
+    // stage 2 fails.
+    for (attempt <- 0 until Stage.MAX_STAGE_FAILURES) {
+      // Complete all the tasks for the current attempt of stage 0 successfully
+      completeNextShuffleMapSuccesfully(0, attempt, numShufflePartitions = 2)
+
+      if (attempt < Stage.MAX_STAGE_FAILURES/2) {
+        // Now we should have a new taskSet, for a new attempt of stage 1.
+        // Fail all these tasks with FetchFailure
+        completeNextStageWithFetchFailure(1, attempt, shuffleDepOne)
+      } else {
+        completeNextShuffleMapSuccesfully(1, attempt, numShufflePartitions = 1)
+
+        // Fail stage 2
+        completeNextStageWithFetchFailure(2, attempt - Stage.MAX_STAGE_FAILURES / 2,
+          shuffleDepTwo)
+      }
+
+      // this will (potentially) trigger a resubmission of stage 0, since we've lost some of its
+      // map output, for the next iteration through the loop
+      scheduler.resubmitFailedStages()
+    }
+
+    completeNextShuffleMapSuccesfully(0, 4, numShufflePartitions = 2)
+    completeNextShuffleMapSuccesfully(1, 4, numShufflePartitions = 1)
+
+    // Succeed stage2 with a "42"
+    completeNextResultStageWithSuccess(2, Stage.MAX_STAGE_FAILURES/2)
+
+    assert(results === Map(0 -> 42))
+    assertDataStructuresEmpty()
+  }
+
+  /**
+   * In this test we simulate a job failure where a stage may have many tasks, many of which fail.
+   * We want to show that many fetch failures inside a single stage attempt do not trigger an abort
+   * on their own, but only when there are enough failing stage attempts.
+   */
+  test("Multiple tasks w/ fetch failures in same stage attempt should not abort the stage.") {
+    setupStageAbortTest(sc)
+
+    val parts = 8
+    val shuffleMapRdd = new MyRDD(sc, parts, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, null)
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, parts, List(shuffleDep))
+    submit(reduceRdd, (0 until parts).toArray)
+
+    completeNextShuffleMapSuccesfully(0, 0, numShufflePartitions = parts)
+
+    completeNextStageWithFetchFailure(1, 0, shuffleDep)
```

Comment:
Can you move this test to be first, since it seems to have a subset of the functionality in the other tests (since all of the other tests also have multiple tasks with fetch failures)? (Or ignore this if you take my suggestion above to  normally only have a single fetch failure)